contrib/adaptive-compression/adapt.c

```diff
@@ -0,0 +1,712 @@
+
+typedef struct {
+    void* start;
+    size_t size;
+    size_t capacity;
+} buffer_t;
+
+typedef struct {
+    size_t filled;
+    buffer_t buffer;
+} inBuff_t;
+
+typedef struct {
+    unsigned waitCompressed;
+    unsigned waitReady;
+    unsigned waitWrite;
+    unsigned readyCounter;
+    unsigned compressedCounter;
+    unsigned writeCounter;
+} cStat_t;
+
+typedef struct {
+    buffer_t src;
+    buffer_t dst;
+    unsigned compressionLevel;
+    unsigned jobID;
+    unsigned lastJob;
+    size_t compressedSize;
+    size_t dictSize;
+} jobDescription;
+
+typedef struct {
+    unsigned compressionLevel;
+    unsigned numActiveThreads;
+    unsigned numJobs;
+    unsigned nextJobID;
+    unsigned threadError;
+    unsigned jobReadyID;
+    unsigned jobCompressedID;
+    unsigned jobWriteID;
+    unsigned allJobsCompleted;
+    unsigned adaptParam;
+    pthread_mutex_t jobCompressed_mutex;
+    pthread_cond_t jobCompressed_cond;
+    pthread_mutex_t jobReady_mutex;
+    pthread_cond_t jobReady_cond;
+    pthread_mutex_t allJobsCompleted_mutex;
+    pthread_cond_t allJobsCompleted_cond;
+    pthread_mutex_t jobWrite_mutex;
+    pthread_cond_t jobWrite_cond;
+    size_t lastDictSize;
+    inBuff_t input;
+    cStat_t stats;
+    jobDescription* jobs;
+    FILE* dstFile;
+    ZSTD_CCtx* cctx;
+} adaptCCtx;
+
+typedef struct {
+    FILE* srcFile;
+    adaptCCtx* ctx;
+} fcResources;
+
+static void freeCompressionJobs(adaptCCtx* ctx)
+{
+    unsigned u;
+    for (u=0; u<ctx->numJobs; u++) {
+        jobDescription job = ctx->jobs[u];
+        free(job.dst.start);
+        free(job.src.start);
+    }
+}
+
+static int freeCCtx(adaptCCtx* ctx)
+{
+    if (!ctx) return 0;
+    {
+        int error = 0;
+        error |= pthread_mutex_destroy(&ctx->jobCompressed_mutex);
+        error |= pthread_cond_destroy(&ctx->jobCompressed_cond);
+        error |= pthread_mutex_destroy(&ctx->jobReady_mutex);
+        error |= pthread_cond_destroy(&ctx->jobReady_cond);
+        error |= pthread_mutex_destroy(&ctx->allJobsCompleted_mutex);
+        error |= pthread_cond_destroy(&ctx->allJobsCompleted_cond);
+        error |= pthread_mutex_destroy(&ctx->jobWrite_mutex);
+        error |= pthread_cond_destroy(&ctx->jobWrite_cond);
+        error |=  (ctx->dstFile != NULL && ctx->dstFile != stdout) ? fclose(ctx->dstFile) : 0;
+        error |= ZSTD_isError(ZSTD_freeCCtx(ctx->cctx));
+        free(ctx->input.buffer.start);
+        if (ctx->jobs){
+            freeCompressionJobs(ctx);
+            free(ctx->jobs);
+        }
+        free(ctx);
+        return error;
+    }
+}
+
+static adaptCCtx* createCCtx(unsigned numJobs, const char* const outFilename)
+{
+
+    adaptCCtx* ctx = calloc(1, sizeof(adaptCCtx));
+    if (ctx == NULL) {
+        DISPLAY("Error: could not allocate space for context\n");
+        return NULL;
+    }
+    ctx->compressionLevel = g_compressionLevel;
+    pthread_mutex_init(&ctx->jobCompressed_mutex, NULL);
```

Comment:
pedantic : 
`pthread_mutex_init()` seems to be able to fail : 
http://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_mutex_init.html

I don't foresee a lot of problems here, but might be worth adding a check at some point.